### PR TITLE
Optimize V1 DP to use tensor collectives for decode inputs & prefill/decode outputs

### PR DESCRIPTION
**Pull request by skhorasganiTT**
- Follow up to https://github.com/tenstorrent/vllm/pull/203.
- Removed usages of `torch.distributed.all_gather_object` in `_execute_model_dp_gather` path for decode inputs and prefill/decode outputs in V1 DP since it was too slow. Previously ~3 ms for input gathers, ~1 ms for output gathers. Now ~0.8-1ms for input gathers, ~0.4 ms for output gathers.
- For decode inputs, added new `build_dp_decode_gather_input` function which tensorizes decode inputs (all padded to same shapes). For block tables, padding to maximum number of blocks in global DP batch instead of absolute max blocks by adding an additional `torch.distributed.all_reduce` (max) collective on the max blocks in each local batch. Input communication across ranks is now done via `torch.distributed.all_gather_into_tensor`.
- For prefill/decode outputs, now padding output tokens to max batch size and using `torch.distributed.all_reduce` (sum) to communicate to ranks (rank 0 sends real data, other ranks send zeros).
- Note: it would be more optimal to use gather instead of all_gather for inputs and scatter instead of all_reduce for outputs, but that is not currently possible since vLLM initializes stateless DP groups which don't support rooted collectives.
- Improves Llama70B-Galaxy V1 decode perf from ~50 t/s/u to ~55 t/s/u, marginal difference in ttft.

vLLM nightly - https://github.com/tenstorrent/tt-metal/actions/runs/18920636984 (note V1 DP tests still missing, CI only tests that V0 functionality and V1 non-DP functionality is intact).

### CODE REVIEW

File: vllm/v1/engine/core.py [hunk 1]
```diff
@@ -1168,45 +1170,74 @@ def _execute_model_dp_gather(self, scheduler_output: SchedulerOutput):
...
+                "concat_and_execute_dp",
+                args=(gathered_inputs, is_decode, max_blocks_decode))[0]
+            assert isinstance(send_tensor, torch.Tensor)
+        else:
+            B = self.vllm_config.scheduler_config.max_num_seqs
+            # Currently only supporting 1 output token per request.
+            send_tensor = torch.zeros((world, B, 1), dtype=torch.int32)
+
+        # Share results via all_gather_into_tensor. Rank 0 contains real data.
+        out = torch.empty((world * world, ) + tuple(send_tensor.shape)[1:],
```
[the remark this answers is not in the document]
Comment (skhorasganiTT) [reply vllm/v1/engine/core.py]: Replace with all reduce, no longer relevant.
